src/cs50/sql.py

```diff
@@ -280,29 +299,18 @@ def execute(self, sql, *args, **kwargs):
             # Infer whether app is defined
             assert flask.current_app
 
-            # If no connection for app's current request yet
-            if not hasattr(flask.g, "_connection"):
-
-                # Connect now
-                flask.g._connection = self._engine.connect()
+            # Disconnect later - but only once
+            if not hasattr(self, "teardown_appcontext_added"):
+                self.teardown_appcontext_added = True
 
-                # Disconnect later
+                # Register shutdown_session on app context teardown
                 @flask.current_app.teardown_appcontext
                 def shutdown_session(exception=None):
-                    if hasattr(flask.g, "_connection"):
-                        flask.g._connection.close()
-
-            # Use this connection
-            connection = flask.g._connection
+                    if self._session is not None:
+                        self._session.close()
 
         except (ModuleNotFoundError, AssertionError):
```

Comment:
Are these ordered based on which might be raised first, given lines above?

Reply:
Not sure what you're asking, so please do clarify if I misunderstood.

The condition checks that `self._session` isn't None because in autocommit mode a transaction is never stored (so the session is already closed, no database connections leaking).

But if there was a manual transaction (meaning the programmer executed `BEGIN`) ongoing then at the end of the Flask request we still want to close the database connection, even if the programmer forgot to commit or rollback.

I also realize that we should probably set `self._session = None` as well when we close the session here, so that will be added in the next commit too.

Let me know if this helps answer your question.

Reply:
Sorry, I just mean why `ModuleNotFoundError, AssertionError` instead of `AssertionError, ModuleNotFoundError`? Just want to ensure it's ordered meaningfully.

Reply:
I didn't modify that line, so not sure for the reasoning there.

Reply:
Ah, nvm, it was me. Is deliberate, because the `import` might raise former and `assert` might raise latter. Aok to leave as is.

Reply:
Great! I'll commit and push updated version in a few minutes.